client/src/main/java/io/pravega/client/stream/Transaction.java

```diff
@@ -70,9 +70,11 @@
      * Send a transaction heartbeat and increase transaction's timeout by lease amount of milliseconds.
      *
      * @param lease Additional amount of time in milliseconds by which to increase transaction's timeout.
+     * @param switchOver Boolean indicating whether the previous successful ping request was performed on a different
+     *                   controller process by the client.
      * @throws PingFailedException Ping failed.
      */
-    void ping(long lease) throws PingFailedException;
+    void ping(long lease, boolean switchOver) throws PingFailedException;
```

Comment:
@skrishnappa 
1. We do maintain txn's timeout information in controllers memory. We do not persist it to stable store in order to avoid accessing stable storage per ping request. However, this is not the only information controller stores in-memory; it also has information about active readers participating in the reader group (scale, commit, abort reader group), which needs to be appropriately handled on controller failure.

2. Secondly, we do things like checking in-memory state to identify whether client is switching over. However, that is not sufficient in the following scenario. (1) Client sends ping to controller C1, (2) Due to n/w glitch it could not access C1 while sending next ping request, so it sends next ping to C2, (3) C2 recognizes that it does not know about the txn, hence it updates stable store to prevent C1 from inadvertently aborting txn (thus, effectively fencing off C1), and starts tracking txn in its in-memory state. (4) Now, due to n/w glitch client cannot access C2, so it sends next ping request again to C1. (5) But C1 may still be actively tracking the txn, so it just updates its in-memory state to increase txn's timeout without fencing off C2, (6) Now after some time C2 can force abort the txn, since it is not fenced off by C1 when it updated timeout for that txn in its memory. Similar situation could arise when client randomly ping requests to arbitrary controller instances.

However, I am now inclined to revert the ping API signature change from this pull request and handle the problem mentioned in 2 above in a separate issue and pull request, because of the following reasons.
* The problem exists even before this pull request, and
* The main goals of this pull request are (a) moving transaction operations out of task framework,  (b) transferring ownership of orphaned transactions on controller failure, and not fixing the problem mentioned in 2 above. 

Note: The way we track transaction timeouts can be thought of as very similar to how ZK tracks session timeouts. ZK server is generally accessed over a local network and ping request is sent every 2 seconds (or 4 seconds). However, in our case client may send ping request over WAN, which is one of the reasons why we allow lease values as high as 30 seconds, so pings need not be sent too frequently. Moreover, once we implement rolling transactions and we get rid of scaleGracePeriod, we can increase permissible lease values to 90-100 seconds or even a higher value depending upon how many concurrent active transactions we intend to support. @shiveshr, sending a ping request in a special stream would also require a RPC call, albeit to SegmentStore instead of controller. We could as well keep ping simple by letting it directly sending RPC request to a controller instance, but let controller instance access stable storage on each ping request. This may be feasible if we make lease value large enough to let controller instance access stable store. For example, if we want to support 10,000 concurrent active transactions, we could live with lease of 100 seconds, as that would translate to 1 ping request on an average per 10 milliseconds and we could easily update ZK node in that interval.